File: lab4/Entities/UserActivity.cs
namespace lab4.Entities
{
    public class UserActivity
    {
        // PK - UserMonth
        public DateTime Month { get; set; } = new DateTime();
        public string UserName { get; set; } = "";
        // PK - Rest
        public int Pid {get; set; } = 0;
        public string ProjectId { get; set; } = "";

        public DateTime Date { get; set; } = new DateTime();
        public string SubactivityId {get; set; } = "";
        public int Time { get; set; } = 0;
        public string? Description { get; set; }

        // Parents
        public virtual UserMonth? UserMonth { get; set; }
        public virtual Subactivity? Subactivity { get; set; }
        public virtual Project? Project { get; set; }
    }
}

File: lab4/ClientApp/src/entities/UserActivity.ts
export class UserActivity {
    month: Date;
    userName: string;
    pid: number;
    projectId: string;
    date: Date;
    subactivityId: string;
    time: number;
    description: string;

    constructor(month: Date, userName: string, pid: number, projectId: string, date: Date, subactivityId: string, time: number, description: string) {
        this.month = month;
        this.userName = userName;
        this.pid = pid;
        this.projectId = projectId;
        this.date = date;
        this.subactivityId = subactivityId;
        this.time = time;
        this.description = description;
    }

    toJSON() {
        return {
            month: this.month,
            userName: this.userName,
            pid: this.pid,
            projectId: this.projectId,
            date: this.date,
            subactivityId: this.subactivityId,
            time: this.time,
            description: this.description
        };
    }

    static fromJSON(json: any) {
        return new UserActivity(json['month'], json['userName'], json['pid'], json['projectId'], json['date'], json['subactivityId'], json['time'], json['description']);
    }

    static createEmpty() {
        return new UserActivity(new Date(), '', 0, '', new Date(), '', 0, '');
    }

    modify(setter: any, key: string, value: any) {
        setter(UserActivity.fromJSON({...this.toJSON(), [key]:value}));
    }
};

export default UserActivity;
